### PR TITLE
loss -- fix embedding

### PR DESCRIPTION
**Pull request by YifanC**
Sean started the first half and I finished the last half of this PR. This branch was forked from an earlier master commit. (Note in the current master aa51757, there's a mismatch in target and to-be-compared data if "shuffle" in the dataloader is on.)

### CODE REVIEW

File: optimize/utils.py [hunk 1]
```diff
@@ -1,6 +1,8 @@
 import numpy as np
 from numpy.lib import recfunctions as rfn
 import torch
+import larndsim.consts 
+import larndsim
```
Comment (sgasioro) [inline optimize/utils.py:5]: These can be removed

File: larndsim/detsim_ep.py [hunk 2]
```diff
@@ -50,12 +51,11 @@ def time_intervals(self, event_id_map, tracks, fields):
         t_start = ep.maximum(ep.full_like(tracks_t_start, self.time_interval[0]),
                              ((tracks_t_start - self.time_padding) / self.t_sampling) * self.t_sampling)
         t_length = t_end - t_start
-        track_starts = (t_start + event_id_map_ep * self.time_interval[1] * 3).raw
+        track_starts = t_start.raw
```
Comment (sgasioro) [inline larndsim/detsim_ep.py:54]: Where is this change coming from?

File: larndsim/detsim_ep.py [hunk 3]
```diff
@@ -413,7 +417,8 @@ def tracks_current(self, pixels, tracks, time_max, fields):
         tpc_borders_ep = ep.from_numpy(pixels, self.tpc_borders).float32()
         borders = ep.stack([tpc_borders_ep[x.astype(int)] for x in tracks_ep[:, fields.index("pixel_plane")]])
 
-        signals = ep.zeros(z_start, shape=(pixels.shape[0], pixels.shape[1], time_max))
+        signals = ep.zeros(z_start, shape=(pixels.shape[0], pixels.shape[1], time_max.astype(int).item()))
+        extras = []
```
Comment (sgasioro) [inline larndsim/detsim_ep.py:421]: I think this was leftover from debugging, should be removed

**Comment by sgasioro** [comment]
Ok, I think this seems mostly fine, will run a couple of tests today to make sure. Some quick thoughts: 
- With the event-wise comparison, we might want to make the data loading a bit more sophisticated, otherwise it'd be very easy to end up with unexpectedly garbage comparisons (e.g. batch is big set of random tracks but we only have one track per event)
- I've been tending to use double precision for everything (I do this in the CI, e.g.), which may need to become the default

File: larndsim/detsim_ep.py (hunk 2 above)
Comment (YifanC) [reply larndsim/detsim_ep.py:54]: If we are calculating loss per event, track_starts won't need to be shifted according to their event id. In another word, `event_id_map_ep` will always be 0. 
In the extreme scenario, when there are tracks from multiple events leaked into the calculation. You'd rather them collapse in the same time window, then have a long scrolling time tick list which would mess up the current setup of the translation from time to drift distance z.

File: larndsim/detsim_ep.py (hunk 3 above)
Comment (YifanC) [reply larndsim/detsim_ep.py]: I removed line 421 `extras = []`. Is there any other residuals?

**Review by YifanC — COMMENTED** [review]
@sgasioro Is this sufficient?
`selected_tracks_torch = selected_tracks_torch.to(self.device).double()`

File: larndsim/detsim_ep.py (hunk 2 above)
Comment (sgasioro) [reply larndsim/detsim_ep.py:54]: Ok, I think this makes sense, and I see you've removed `event_id_map` from the arguments, so that's consistent

File: larndsim/detsim_ep.py (hunk 3 above)
Comment (sgasioro) [reply larndsim/detsim_ep.py]: This was the only one I saw, I think!

**Comment by sgasioro** [comment]
> @sgasioro Is this sufficient? line 132 in optimize/fit_params.py `selected_tracks_torch = selected_tracks_torch.to(self.device).double()`

This works, but it's a bit buried -- I would maybe prefer putting it in `dataio.py`, possibly just defining `__getitem__()` as
```
def __getitem__(self, idx):
    return self.tracks[idx].double()
```

Does it fix the reproducibility issues you were having?

**Review by sgasioro — APPROVED** [review]
Also yeah, some quick looks this loss function seems fine, though maybe some remaining gradient things to sort out -- but I'm happy to have this merged
